### PR TITLE
fix(crypto) fix KeyObjectkey handling

### PR DESCRIPTION
**Pull request by cirospaciari**
### What does this PR do?

- [ ] Documentation or TypeScript types (it's okay to leave the rest blank in this case)
- [x] Code changes

### How did you verify your code works?

Existing tests

### CODE REVIEW

File: src/bun.js/bindings/KeyObject.cpp [hunk 1]
```diff
@@ -130,53 +130,119 @@ struct AsymmetricKeyValueWithDER {
 
 class KeyPassphrase {
 private:
-    WTF::CString pass_holder;
+    union {
+        WTF::CStringBuffer* strBuffer;
+        JSUint8Array* array;
+    } u;
+    uint8_t tag;
```
Comment (gvilums) [inline src/bun.js/bindings/KeyObject.cpp]: I think this should be an enum with a default value to make it clear what each tag value is referring to, e.g.
```cpp
enum class Tag {
  None,
  StrBuffer,
  Array,
}
```
We could then add an initializer:
`Tag tag = Tag::None;`

File: src/bun.js/bindings/KeyObject.cpp [hunk 2]
```diff
@@ -130,53 +130,119 @@ struct AsymmetricKeyValueWithDER {
...
+    {
+        return tag > 0;
+    }
+
+    char* data()
+    {
+        switch (tag) {
+        case 1: {
+            if (this->u.strBuffer) {
+                return const_cast<char*>(this->u.strBuffer->data());
```
Comment (gvilums) [inline src/bun.js/bindings/KeyObject.cpp]: Is this `const_cast` safe? If we were to write to it and the underlying data was const we'd introduce UB

File: src/bun.js/bindings/KeyObject.cpp [hunk 3]
```diff
@@ -130,53 +130,119 @@ struct AsymmetricKeyValueWithDER {
...
+            }
+            return nullptr;
+        }
+        case 2: {
+            if (this->u.array) {
+                return (char*)(this->u.array->vector());
```
Comment (gvilums) [inline src/bun.js/bindings/KeyObject.cpp]: Why a C-style cast here?

**Review by gvilums — CHANGES_REQUESTED** [review]

File: src/bun.js/bindings/KeyObject.cpp (hunk 3 above)
Comment (cirospaciari) [reply src/bun.js/bindings/KeyObject.cpp]: Is just a void* casted to char* I think it already was using C style I will change that

File: src/bun.js/bindings/KeyObject.cpp (hunk 2 above)
Comment (cirospaciari) [reply src/bun.js/bindings/KeyObject.cpp]: Is safe internally is actually mutable data, and the data is not actually mutate in any manner anyway

**Comment by Jarred-Sumner** [comment]
From this diff, it’s not clear to me why it was crashing before - which means it’s not clear why this fixes it. Which line was the fix?

**Comment by cirospaciari** [comment]
> From this diff, it’s not clear to me why it was crashing before - which means it’s not clear why this fixes it. Which line was the fix?

buffer->ref() and buffer->deref() to keep it alive was the fix (increasing the buffer ref count and decreasing when going out of scope)

**Pull request merged**